Use multiprocessing for the scheduler

bolkedebruin opened the pull request:
This parallelizes the scheduler to the amount of cpu-cores.

criccomini commented on airflow/jobs.py:
```diff
@@ -422,7 +424,7 @@ def schedule_dag(self, dag):
             # don't ever schedule prior to the dag's start_date
             if dag.start_date:
-                next_run_date = max(next_run_date, dag.start_date)
+                next_run_date = max(next_run_date or dag.start_date, dag.start_date)
```
Nice catch. Perhaps a little more intelligible (though more verbose) would be:

```
dag.start_date if not next_run_date else max(next_run_date, dag.start_date)
```

criccomini commented on airflow/jobs.py:
```diff
@@ -657,6 +659,23 @@ def prioritize_queued(self, session, executor, dagbag):
 
                 session.commit()
 
+    def _split_dags(self, dags, size):
```
Pydocs would be helpful. This stuff is getting a little gnarly.

criccomini commented on airflow/jobs.py:
```diff
@@ -657,6 +659,23 @@ def prioritize_queued(self, session, executor, dagbag):
+        size = max(1, size)
+        return [dags[i:i + size] for i in range(0, len(dags), size)]
+
+    def _do_dags(self, dagbag, dags):
```
Pydocs would be helpful. This stuff is getting a little gnarly.

criccomini commented on airflow/jobs.py:
```diff
@@ -695,18 +714,19 @@ def _execute(self):
-                        self.manage_slas(dag)
-                    except Exception as e:
-                        self.logger.exception(e)
+                dags = [x for x in dags if x.dag_id not in paused_dag_ids]
```
I think you can use [filter()](https://docs.python.org/2/library/functions.html#filter) to this a little more pythonically.

bolkedebruin replied:
with a lambda? I personally don't like to read that stuff so much

criccomini commented on airflow/jobs.py:
```diff
@@ -695,18 +714,19 @@ def _execute(self):
-                    except Exception as e:
-                        self.logger.exception(e)
+                dags = [x for x in dags if x.dag_id not in paused_dag_ids]
+                self.logger.info("CPU: {} DAGs:{}".format(multiprocessing.cpu_count(), len(dags)))
```
Nit: "Total CPUs: {}" would be a little more intelligible. At first, I thought this was saying "I am CPU number N, I'm going to process DAGs 1-10"

criccomini replied:
How expensive is cpu_count? I notice we're calling it multiple times in the scheduler loop.

bolkedebruin replied:
don't know to be honest. I haven't seen any impact though

criccomini commented on airflow/utils/db.py:
```diff
@@ -83,6 +83,23 @@ def merge_conn(conn, session=None):
         session.commit()
 
 
+@event.listens_for(settings.engine, "connect")
```
What are these? MySQLAlchemy foo is lacking.

bolkedebruin replied:
It basically checks when a new connecting is made. You cannot use the same db connection in a fork so we force the db logic to make a new connection when it detects that the pid of the process that created the connection is not the same as it is currently in.

r39132 commented on airflow/configuration.py:
```diff
@@ -414,6 +420,7 @@ def run_command(command):
 job_heartbeat_sec = 1
 scheduler_heartbeat_sec = 5
 authenticate = true
+max_threads = 2
```
Nit  : maybe we should call this max_proc?

bolkedebruin replied:
I was thinking "threads" is better understood as a concept. The executor uses `parallelism` but that is also a bit unclear.

r39132 commented on airflow/jobs.py:
```diff
@@ -422,7 +435,7 @@ def schedule_dag(self, dag):
             # don't ever schedule prior to the dag's start_date
             if dag.start_date:
-                next_run_date = max(next_run_date, dag.start_date)
+                next_run_date = dag.start_date if not next_run_date else max(next_run_date, dag.start_date)
```
This is an important bug fix :-) Should be a stand-alone commit or PR.

It looks like the cause of this is that a `TaskInstance` has not yet been inserted for the first run. As a result, even though `last_scheduled_run` is `None`, then `latest_run` is also `None`.  If `latest_run` is `None`, then the logic above picks the min start date for all of the tasks in the dag. This should never be None -- this is where there might be a bug. 

The logic flow above, for others looking at this code for the first time is 
1) look at dagruns (which the scheduler inserts into - `airflow run` and `airflow backfill` do not currently insert into this table) 
2) look at task_instance (which is presumably inserted after dag_runs) 
3) look at the dag bag for a start date

It seems counterintuitive that the tasks in the dagbag would not have a start date

bolkedebruin replied:
Ok it is a separate commit now. Real fix should be in another PR I guess.

jlowin commented on airflow/jobs.py:
```diff
@@ -668,8 +706,8 @@ def _execute(self):
         executor.start()
-        i = 0
-        while not self.num_runs or self.num_runs > i:
+        self.runs = 0
```
Is there an advantage to making this `self.runs` instead of just `runs`?

bolkedebruin replied:
Yes, that it actually counts :-).  When I kept it a local variable it would not properly increase. I am not sure why.

bolkedebruin committed:
Partial fix to make sure next_run_date cannot be None

It seems possible that a TaskInstance has not yet been inserted for the first run.
As a result, even though last_scheduled_run is None, then latest_run is also None.
If latest_run is None, then the logic above picks the min start date for all of the
tasks in the dag. This should never be None -- this is where there might be a bug.

This patch works around the issue by making sure a next_run_date is always set to
either the dag start date if the TaskInstance has no calculated next_run_date yet.

bolkedebruin committed:
Add multiprocessing support to the scheduler

As the amount of dags grows and the ability to create dags programmatically
is more often used, more and more time is spend in the scheduler which lowers
throughput.

This patch adds the ability to use multiple threads to the scheduler. The
amount of threads can be specified by "max_threads" in the scheduler section
of the configuration. The amount of threads will, however, not exceed the
amount of cores.

In case of using sqlite the max_threads will be set to 1 as sqlite does not
support multiple db connections.
